docs/html/reference/pip_install.rst

```diff
@@ -780,6 +780,25 @@ No other build system commands are invoked by the ``pip install`` command.
 
 Installing a package from a wheel does not invoke the build system at all.
 
+Dependency-only installations
++++++++++++++++++++++++++++++
+
+In some situations it can be useful to only install the dependencies of a
+package and avoid installing the package itself. An example would be setting up
+a development environment with all development dependencies but not the
+package itself. This can be done with the `--no-deps` option:
+
+$ pip install --deps-only .[test]
+
+Note that there can be cases when not all dependencies will be installed when
+using this option. Consider an example where package A is a dependency of B
+and the following command is issued:
+
+$ pip install --deps-only A B
+
+Pip will skip the installation of A and B, but because B depends on A a
+dependency won't be satisfied.
```

Comment:
Documenting this doesn't make it right - in fact, reading the docs makes me even less convinced that the behaviour you've documented here would be what people want. Also, IMO, you need to document the fact that the above command behaves *differently* from
```
$ pip install --deps-only A
$ pip install --deps-only B
```

And honestly, I'm still not happy that discrepancy exists. It seems like a recipe for people to make mistakes.

My feeling is that you're focusing too closely on how *you* want to use the option, and not thinking about other ways people might use it (either deliberately or in error). As I don't personally see a use case for the option (I don't see why `pip install -e .` doesn't work for you, it works for all the cases I've ever needed) I'm tending to focus more on how people might make mistakes with it - I'm not trying to tear your suggestion apart, I just have a different perspective :-)

Reply:
Thanks for having a look @pfmoore. I agree that the behaviour can be surprising and I'm not 100% sure either if this is the right way of doing it.

Let me describe one use case where this would be necessary, perhaps you can suggest a better way of doing it. 

I want to use `extras_require` to maintain extra dependencies in my libraries for testing, building the documentation, etc. as recommended [here](https://snarky.ca/clarifying-pep-518/). The problem with this is that when it comes to installing and testing the `sdist` of said library I have no way of installing only the test dependencies without installing the package itself. Installing the package in develop mode does not test the final "artifact" so a bad `sdist` can slip through testing. `--deps-only` might not be the right way of fixing this but I'm not sure how else this use can could be supported.

Reply:
Well, I'm not sure I agree with the recommendation to use `extras_require` - I'd just use a requirements file. So I don't really have an answer for you in that case.

Maybe this *is* the right way of doing this. But just because there's a use case for installing "only the dependencies but not the main package" doesn't mean that the inconsistencies I pointed out are acceptable. I'm not rejecting the PR, nor even saying that the inconsistencies *have* to be solved - just that I'm not comfortable myself with merging the PR myself because of the potential maintenance issues. One of the other pip devs may be willing to take a more relaxed view of the issues.